### ISSUE TITLE
DiffusionPolicy: no EMA support — the reference implementation enables it in every config and evaluates the EMA weights

### ISSUE THREAD

**Issue by 2047767028-lang** [huggingface/lerobot#4259]
### Summary

lerobot's Diffusion Policy has no EMA (exponential moving average of model weights) anywhere — no implementation, no config flag. The original implementation ([real-stanford/diffusion_policy](https://github.com/real-stanford/diffusion_policy)) treats EMA as a default part of the method: **every shipped config enables it, and the real-robot evaluation script loads the EMA weights, not the raw ones.** Users porting the official hyperparameters to lerobot are therefore training under different dynamics than the ones those hyperparameters were tuned for.

### What is missing (checked against `main`, 2026-07-31, and 0.3.3)

- `src/lerobot/policies/diffusion/configuration_diffusion.py` — no EMA mention, no flag
- `src/lerobot/policies/diffusion/modeling_diffusion.py` — no EMA mention
- `src/lerobot/scripts/lerobot_train.py` — training loop updates the policy weights directly every step; no shadow weights

(Same result on 0.3.3: full grep of `policies/diffusion/`, `scripts/train.py`, and the `TrainPipelineConfig` fields.)

### What the reference implementation does

`diffusion_policy/config/train_diffusion_unet_image_workspace.yaml` (sim benchmark) — verbatim:

```yaml
use_ema: True
ema:
  _target_: diffusion_policy.model.diffusion.ema_model.EMAModel
  update_after_step: 0
  inv_gamma: 1.0
  power: 0.75
  min_value: 0.0
  max_value: 0.9999
```

`train_diffusion_unet_real_image_workspace.yaml` (real robot) likewise sets `use_ema: True`.

And `eval_real_robot.py` evaluates the EMA model, not the raw weights:

```python
if cfg.training.use_ema:
    policy = workspace.ema_model
```

### Why it matters

- **Reproducibility**: anyone copying the official hyperparameters into lerobot silently drops a component the authors ship by default in every config.
- **Small-batch users are hit hardest.** The official batch size is 64. Users on small GPUs train at batch 8–16 with correspondingly noisier gradients — exactly the regime where weight averaging does the most work. In the official stack EMA absorbs that noise; in lerobot there is nothing.
- Diffusion outputs are consumed directly as robot actions, so weight jitter translates into action jitter at deployment.

### Implementation path

lerobot already depends on `diffusers` (`modeling_diffusion.py` imports its DDPM/DDIM schedulers), and `diffusers.training_utils.EMAModel` is available in the same environment — so an implementation would not add a dependency. The design question is where it lives (policy-level vs. trainer-level), which affects other policies too; happy to contribute a PR if maintainers have a preferred direction.

### Workaround in the meantime

Post-hoc checkpoint averaging (SWA): with a cosine schedule the trailing checkpoints sit in the same basin, and averaging the last K of them approximates what EMA would have produced — zero library changes. A ready-made script (including a `--report-only` mode that first measures whether your trailing checkpoints differ enough for averaging to matter) is here: https://github.com/2047767028-lang/lerobot-dp-toolkit

### Minor related note

`DiffusionConfig.get_optimizer_preset()` returns `AdamConfig` (plain Adam), while the reference implementation uses `torch.optim.AdamW` (`weight_decay: 1.0e-6`, `betas: [0.95, 0.999]`). At that weight decay the difference is likely negligible, but strictly speaking it is one more deviation from the reference recipe.

**Comment by alejodosr** [comment]
I'd like to pick this up working from today on

EDIT: 

I'd like to work on this. I have a working implementation locally, planning to send a PR soon. The approach:

- EMA at the trainer level, opt-in via `--ema.enable=true` (new `EMAConfig` on `TrainPipelineConfig`), off by default so nothing changes for existing users
- uses `diffusers.EMAModel` with the same warmup schedule as the reference implementation (`inv_gamma=1.0, power=0.75`). Imported lazily, so no new dependency (the diffusion policy already needs diffusers anyway)
- EMA state saved with checkpoints for exact resume, plus a loadable `pretrained_model_ema/` copy next to the regular one so you can `eval/deploy` the EMA weights directly
- when enabled, periodic env eval runs on the EMA weights, and the final push also uploads them to a sibling `<repo_id>-ema` repo

@pkooij before I open it: I saw you had a pretty similar trainer-level EMA on a branch and removed it in `147b8f2`. Was that a decision against having EMA in the training pipeline, or just cleanup of that branch? Don't want to send a PR against something that was already decided.

**Comment by alejodosr** [comment]
@2047767028-lang feel free to try the PR, to give feedback on your use case

**Comment by 2047767028-lang** [comment]
thank for your help, i will try it to train a better DP model
